pkg/api/util/ipc_endpoint.go

```diff
@@ -0,0 +1,144 @@
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2023-present Datadog, Inc.
+
+package util
+
+import (
+	"crypto/tls"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+	"strconv"
+
+	"github.com/DataDog/datadog-agent/comp/core/config"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	"github.com/DataDog/datadog-agent/pkg/util/system"
+)
+
+// IPCEndpoint is an endpoint that IPC requests will be sent to
+type IPCEndpoint struct {
+	client    *http.Client
+	target    url.URL
+	closeConn bool
+}
+
+// GetOption is an option that can be passed to DoGet
+type GetOption func(url.URL) url.URL
+
+// WithValues is an option to add url.Values to the GET request
+func WithValues(values url.Values) GetOption {
+	return func(u url.URL) url.URL {
+		u.RawQuery = values.Encode()
+		return u
+	}
+}
+
+// DoGet sends GET method to the endpoint
+func (end *IPCEndpoint) DoGet(options ...GetOption) ([]byte, error) {
+	conn := LeaveConnectionOpen
+	if end.closeConn {
+		conn = CloseConnection
+	}
+
+	target := end.target
+	for _, opt := range options {
+		target = opt(target)
+	}
+
+	// TODO: after removing callers to api/util/DoGet, pass `end.token` instead of using global var
+	res, err := DoGet(end.client, target.String(), conn)
+	if err != nil {
+		var errMap = make(map[string]string)
+		_ = json.Unmarshal(res, &errMap) //nolint:errcheck
+		// If the error has been marshalled into a json object, check it and return it properly
+		if errStr, found := errMap["error"]; found {
+			return nil, errors.New(errStr)
+		}
+
+		return nil, fmt.Errorf("could not reach agent: %v\nMake sure the agent is running before requesting the runtime configuration and contact support if you continue having issues", err)
+	}
+	return res, err
+}
+
+// EndpointOption allows configuration of the IPCEndpoint during construction
+type EndpointOption func(*IPCEndpoint)
+
+// WithCloseConnection is an option to close the connection
+func WithCloseConnection(state bool) func(*IPCEndpoint) {
+	return func(end *IPCEndpoint) {
+		end.closeConn = state
+	}
+}
+
+// WithHTTPClient is an option to assign a different http.Client
+func WithHTTPClient(client *http.Client) func(*IPCEndpoint) {
+	return func(end *IPCEndpoint) {
+		end.client = client
+	}
+}
+
+// WithURLScheme is an option to set the URL's scheme
+func WithURLScheme(scheme string) func(*IPCEndpoint) {
+	return func(end *IPCEndpoint) {
+		end.target.Scheme = scheme
+	}
+}
+
+// WithHostAndPort is an option to use a host address for sending IPC requests
+func WithHostAndPort(cmdHost string, cmdPort int) func(*IPCEndpoint) {
+	return func(end *IPCEndpoint) {
+		end.target.Host = net.JoinHostPort(cmdHost, strconv.Itoa(cmdPort))
+	}
+}
+
+// NewIPCEndpoint constructs a new IPC Endpoint using the given config, path, and options
+func NewIPCEndpoint(config config.Component, endpointPath string, options ...EndpointOption) (*IPCEndpoint, error) {
```

Comment:
I'm not super happy with doing so: as mentioned in the jira issue, the main motivation is removing references to the global `pkgconfig.Datadog` object. Requiring the `config.Component` interface accomplishes this, but using `model.Reader` means callers do not actually have to stop using that global.

So far, we've been keeping the root folder of each component light on dependencies, usually they contain just the component's interface, and we keep the implementation in a subfolder. I realize that is hard to do for the config because it's barely componentized. But would it be possible to move the implementation from `config_sync.go` into another location using fx?

That may be too hard to do just yet, and I don't want to block your work, so I could make this change for now. But I'd want to change this param back to `config.Component` soon and better hide the details of the config construction into a subfolder to avoid this cycle. What do you think?

Reply:
> the main motivation is removing references to the global pkgconfig.Datadog object. Requiring the config.Component interface accomplishes this, but using model.Reader means callers do not actually have to stop using that global.

That's true, but at the same time it means that if you don't have the config component yet then you just can't use the util functions... While you actually don't need to take the component, you only need a `config.Reader`

> we've been keeping the root folder of each component light on dependencies, usually they contain just the component's interface, and we keep the implementation in a subfolder.

That would indeed solve the import cycle in that case, and it shouldn't be super hard to do, I'll take a look tomorrow 👍 

I might move my config sync logic in a separate component to fix some other issue so this conflict might just disappear anyway 🤞

Reply:
> That's true, but at the same time it means that if you don't have the config component yet then you just can't use the util functions

Check out the list in ASCII-1186. All of the agent subcommands have the `config.Compoment` available, but they're using `pkgconfig.Datadog` currently.

Some locations like `pkg/diagnose/runner.go` don't have the config, but I think we should fix them if possible in order to remove their direct use of `pkgconfig.Datadog`.

> While you actually don't need to take the component, you only need a config.Reader

On the one hand, there's the minimum amount the code needs to do to function, on the other hand, there's best practices for future maintainability. References to global state falls into the first category, and I would argue `pkgconfig.Reader` does as well. Yes, it works, but it leads to issues down the line. By having stricter requirements, we can move the code closer to the second category.

> I might move my config sync logic in a separate component to fix some other issue so this conflict might just disappear anyway

If that works, that would be excellent, it would solve both your issues and mine.